### PR TITLE
ROCm contributions week 28

### PR DESCRIPTION
**Pull request by iotamudelta**
This PR contains the ROCm contributions of last week:
* documentation of pyHIPIFY data format originating from #8812 reviewing comments by @ezyang 
* removal of most patch files from the `amd_build` directory and integration into the code base
* enabling of previously disabled_features that do compile now
* improvement to the static_cast feature in pyHIPIFY (it will only apply static_cast to kernel arguments, not launch arguments)
* addition of two workarounds to pyHIPIFY for ROCm/HIP shortcomings: a) `__forceinline__` does not imply `static`, hence change to `__inline__`, b) `std::[exp,log,pow]` math functions cannot be selected in device code, use `::[exp,log,pow]` instead. Both of these workarounds will be removed once the issues are fixed upstream. Neither of these issues have surfaced on the CI but were reproduced internally.

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -118,8 +118,12 @@ void SpatialSoftMax_getLaunchSizes(
   uint32_t block_threads = block.x * block.y;
   smem_size = block.x == 1 ? 0 : block_threads * sizeof(accscalar_t);
   int max_active_blocks;
+#ifdef __HIP_PLATFORM_HCC__
+  max_active_blocks = 16;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/SoftMax.cu]: Where is this HIP limitation documented?

File: aten/src/THC/generic/THCTensorRandom.cu [hunk 2]
```diff
@@ -490,11 +490,11 @@ THC_API void THCTensor_(clampedRandom)(THCState* state, THCTensor *self_, int64_
 #if defined(THC_REAL_IS_LONG) || defined(THC_REAL_IS_DOUBLE) || defined(THC_REAL_IS_FLOAT)
   if (range > 1ULL << 32) {
     generate_random_64<<<NUM_BLOCKS, BLOCK_SIZE, 0, THCState_getCurrentStream(state)>>>(
-        gen->state.gen_states, size, data, min_val, range);
+        gen->state.gen_states, static_cast<int32_t>(size), data, min_val, range);
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorRandom.cu]: Why isn't a `static_cast<int>` sufficient here? The signature of the macroed function is:

```
__global__ void NAME(curandStateMtgp32 *state, int size, T *result, ARG1, ARG2)      \
```

Certainly on many platforms `int` is 32-bit, but I wouldn't rely on that.

File: tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py [hunk 3]
```diff
@@ -1,5 +1,17 @@
...
+""" Mapping of CUDA functions, include files, constants, and types to ROCm/HIP equivalents
+
+This closely follows the implementation in hipify-clang
+https://github.com/ROCm-Developer-Tools/HIP/blob/master/hipify-clang/src/CUDA2HipMap.cpp
+and its structure.
+There are different maps for fundamental names, include files, identifies, sparse, and
+PyTorch specific translations.
+Each of the entries in these maps translates a CUDA string to a tuple containing the
+ROCm/HIP string, a type and API annotation and - optionally - an annotation if it is not
+supported in ROCm/HIP yet.
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py]: Thank you, these comments help a lot.

**Review by ezyang — APPROVED** [review]
Looks basically reasonable.

**Comment by ezyang** [comment]
@iotamudelta Do you mind fixing those minor nits? Then I can merge it.

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (iotamudelta) [reply aten/src/ATen/native/cuda/SoftMax.cu]: it isn't AFAIK, we are working on that.

File: aten/src/THC/generic/THCTensorRandom.cu (hunk 2 above)
Comment (iotamudelta) [reply aten/src/THC/generic/THCTensorRandom.cu]: Yeah, that should do. I'll change that.

**Pull request closed without merging**